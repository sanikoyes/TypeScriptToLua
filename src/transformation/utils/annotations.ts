import * as ts from "typescript";
import { TransformationContext } from "../context";

export enum AnnotationKind {
    Extension = "extension",
    MetaExtension = "metaExtension",
    CustomConstructor = "customConstructor",
    CompileMembersOnly = "compileMembersOnly",
    NoResolution = "noResolution",
    PureAbstract = "pureAbstract",
    Phantom = "phantom",
    TupleReturn = "tupleReturn",
    LuaIterator = "luaIterator",
    LuaTable = "luaTable",
    NoSelf = "noSelf",
    NoSelfInFile = "noSelfInFile",
    Vararg = "vararg",
    ForRange = "forRange",
    Jsx = "jsx",
    JsxFrag = "jsxFrag",
}

export interface Annotation {
    kind: AnnotationKind;
    args: string[];
}

function createAnnotation(name: string, args: string[]): Annotation | undefined {
    const kind = Object.values(AnnotationKind).find(k => k.toLowerCase() === name.toLowerCase());
    if (kind !== undefined) {
        return { kind, args };
    }
}

export type AnnotationsMap = Map<AnnotationKind, Annotation>;

function collectAnnotations(source: ts.Symbol | ts.Signature, annotationsMap: AnnotationsMap): void {
    for (const tag of source.getJsDocTags()) {
        const annotation = createAnnotation(tag.name, tag.text?.map(p => p.text) ?? []);
        if (annotation) {
            annotationsMap.set(annotation.kind, annotation);
        }
    }
}

export function getSymbolAnnotations(symbol: ts.Symbol): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();
    collectAnnotations(symbol, annotationsMap);
    return annotationsMap;
}

export function getTypeAnnotations(type: ts.Type): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    if (type.symbol) collectAnnotations(type.symbol, annotationsMap);
    if (type.aliasSymbol) collectAnnotations(type.aliasSymbol, annotationsMap);

    return annotationsMap;
}

export function getNodeAnnotations(node: ts.Node): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    for (const tag of ts.getJSDocTags(node)) {
        const tagName = tag.tagName.text;
        const annotation = createAnnotation(tagName, getTagArgsFromComment(tag));
        if (annotation) {
            annotationsMap.set(annotation.kind, annotation);
        }
    }

    return annotationsMap;
}

export function getFileAnnotations(sourceFile: ts.SourceFile): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    if (sourceFile.statements.length > 0) {
        // Manually collect jsDoc because `getJSDocTags` includes tags only from closest comment
        const jsDoc = sourceFile.statements[0].jsDoc;
        if (jsDoc) {
            for (const tag of jsDoc.flatMap(x => x.tags ?? [])) {
                const tagName = tag.tagName.text;
                const annotation = createAnnotation(tagName, getTagArgsFromComment(tag));
                if (annotation) {
                    annotationsMap.set(annotation.kind, annotation);
                }
            }
        }
    }

    return annotationsMap;
}

export function getSignatureAnnotations(context: TransformationContext, signature: ts.Signature): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();
    collectAnnotations(signature, annotationsMap);

    // Function properties on interfaces have the JSDoc tags on the parent PropertySignature
    const declaration = signature.getDeclaration();
    if (declaration?.parent && ts.isPropertySignature(declaration.parent)) {
        const symbol = context.checker.getSymbolAtLocation(declaration.parent.name);
        if (symbol) {
            collectAnnotations(symbol, annotationsMap);
        }
    }

    return annotationsMap;
}

export function isTupleReturnCall(context: TransformationContext, node: ts.Node): boolean {
    if (!ts.isCallExpression(node)) {
        return false;
    }

    const signature = context.checker.getResolvedSignature(node);
    if (signature) {
        if (getSignatureAnnotations(context, signature).has(AnnotationKind.TupleReturn)) {
            return true;
        }

        // Only check function type for directive if it is declared as an interface or type alias
        const declaration = signature.getDeclaration();
        const isInterfaceOrAlias =
            declaration?.parent &&
            ((ts.isInterfaceDeclaration(declaration.parent) && ts.isCallSignatureDeclaration(declaration)) ||
                ts.isTypeAliasDeclaration(declaration.parent));
        if (!isInterfaceOrAlias) {
            return false;
        }
    }

    const type = context.checker.getTypeAtLocation(node.expression);
    return getTypeAnnotations(type).has(AnnotationKind.TupleReturn);
}

export function isLuaIteratorType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.LuaIterator);
}

export function isVarargType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.Vararg);
}

export function isForRangeType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.ForRange);
}

export function getTagArgsFromComment(tag: ts.JSDocTag): string[] {
    if (tag.comment) {
        if (typeof tag.comment === "string") {
            return tag.comment.split(" ");
        } else {
            return tag.comment.map(part => part.text);
        }
    }

    return [];
}
