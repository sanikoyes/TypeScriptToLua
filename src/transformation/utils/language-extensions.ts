import * as ts from "typescript";
import { TransformationContext } from "../context";

export enum ExtensionKind {
    MultiFunction = "MultiFunction",
    MultiType = "MultiType",
    RangeFunction = "RangeFunction",
    VarargConstant = "VarargConstant",
    IterableType = "IterableType",
    AdditionOperatorType = "AdditionOperatorType",
    AdditionOperatorMethodType = "AdditionOperatorMethodType",
    SubtractionOperatorType = "SubtractionOperatorType",
    SubtractionOperatorMethodType = "SubtractionOperatorMethodType",
    MultiplicationOperatorType = "MultiplicationOperatorType",
    MultiplicationOperatorMethodType = "MultiplicationOperatorMethodType",
    DivisionOperatorType = "DivisionOperatorType",
    DivisionOperatorMethodType = "DivisionOperatorMethodType",
    ModuloOperatorType = "ModuloOperatorType",
    ModuloOperatorMethodType = "ModuloOperatorMethodType",
    PowerOperatorType = "PowerOperatorType",
    PowerOperatorMethodType = "PowerOperatorMethodType",
    FloorDivisionOperatorType = "FloorDivisionOperatorType",
    FloorDivisionOperatorMethodType = "FloorDivisionOperatorMethodType",
    BitwiseAndOperatorType = "BitwiseAndOperatorType",
    BitwiseAndOperatorMethodType = "BitwiseAndOperatorMethodType",
    BitwiseOrOperatorType = "BitwiseOrOperatorType",
    BitwiseOrOperatorMethodType = "BitwiseOrOperatorMethodType",
    BitwiseExclusiveOrOperatorType = "BitwiseExclusiveOrOperatorType",
    BitwiseExclusiveOrOperatorMethodType = "BitwiseExclusiveOrOperatorMethodType",
    BitwiseLeftShiftOperatorType = "BitwiseLeftShiftOperatorType",
    BitwiseLeftShiftOperatorMethodType = "BitwiseLeftShiftOperatorMethodType",
    BitwiseRightShiftOperatorType = "BitwiseRightShiftOperatorType",
    BitwiseRightShiftOperatorMethodType = "BitwiseRightShiftOperatorMethodType",
    ConcatOperatorType = "ConcatOperatorType",
    ConcatOperatorMethodType = "ConcatOperatorMethodType",
    LessThanOperatorType = "LessThanOperatorType",
    LessThanOperatorMethodType = "LessThanOperatorMethodType",
    GreaterThanOperatorType = "GreaterThanOperatorType",
    GreaterThanOperatorMethodType = "GreaterThanOperatorMethodType",
    NegationOperatorType = "NegationOperatorType",
    NegationOperatorMethodType = "NegationOperatorMethodType",
    BitwiseNotOperatorType = "BitwiseNotOperatorType",
    BitwiseNotOperatorMethodType = "BitwiseNotOperatorMethodType",
    LengthOperatorType = "LengthOperatorType",
    LengthOperatorMethodType = "LengthOperatorMethodType",
    TableNewType = "TableNewType",
    TableDeleteType = "TableDeleteType",
    TableDeleteMethodType = "TableDeleteMethodType",
    TableGetType = "TableGetType",
    TableGetMethodType = "TableGetMethodType",
    TableHasType = "TableHasType",
    TableHasMethodType = "TableHasMethodType",
    TableSetType = "TableSetType",
    TableSetMethodType = "TableSetMethodType",
}

const extensionKindToValueName: { [T in ExtensionKind]?: string } = {
    [ExtensionKind.MultiFunction]: "$multi",
    [ExtensionKind.RangeFunction]: "$range",
    [ExtensionKind.VarargConstant]: "$vararg",
};

const extensionKindToTypeBrand: { [T in ExtensionKind]: string } = {
    [ExtensionKind.MultiFunction]: "__luaMultiFunctionBrand",
    [ExtensionKind.MultiType]: "__luaMultiReturnBrand",
    [ExtensionKind.RangeFunction]: "__luaRangeFunctionBrand",
    [ExtensionKind.VarargConstant]: "__luaVarargConstantBrand",
    [ExtensionKind.IterableType]: "__luaIterableBrand",
    [ExtensionKind.AdditionOperatorType]: "__luaAdditionBrand",
    [ExtensionKind.AdditionOperatorMethodType]: "__luaAdditionMethodBrand",
    [ExtensionKind.SubtractionOperatorType]: "__luaSubtractionBrand",
    [ExtensionKind.SubtractionOperatorMethodType]: "__luaSubtractionMethodBrand",
    [ExtensionKind.MultiplicationOperatorType]: "__luaMultiplicationBrand",
    [ExtensionKind.MultiplicationOperatorMethodType]: "__luaMultiplicationMethodBrand",
    [ExtensionKind.DivisionOperatorType]: "__luaDivisionBrand",
    [ExtensionKind.DivisionOperatorMethodType]: "__luaDivisionMethodBrand",
    [ExtensionKind.ModuloOperatorType]: "__luaModuloBrand",
    [ExtensionKind.ModuloOperatorMethodType]: "__luaModuloMethodBrand",
    [ExtensionKind.PowerOperatorType]: "__luaPowerBrand",
    [ExtensionKind.PowerOperatorMethodType]: "__luaPowerMethodBrand",
    [ExtensionKind.FloorDivisionOperatorType]: "__luaFloorDivisionBrand",
    [ExtensionKind.FloorDivisionOperatorMethodType]: "__luaFloorDivisionMethodBrand",
    [ExtensionKind.BitwiseAndOperatorType]: "__luaBitwiseAndBrand",
    [ExtensionKind.BitwiseAndOperatorMethodType]: "__luaBitwiseAndMethodBrand",
    [ExtensionKind.BitwiseOrOperatorType]: "__luaBitwiseOrBrand",
    [ExtensionKind.BitwiseOrOperatorMethodType]: "__luaBitwiseOrMethodBrand",
    [ExtensionKind.BitwiseExclusiveOrOperatorType]: "__luaBitwiseExclusiveOrBrand",
    [ExtensionKind.BitwiseExclusiveOrOperatorMethodType]: "__luaBitwiseExclusiveOrMethodBrand",
    [ExtensionKind.BitwiseLeftShiftOperatorType]: "__luaBitwiseLeftShiftBrand",
    [ExtensionKind.BitwiseLeftShiftOperatorMethodType]: "__luaBitwiseLeftShiftMethodBrand",
    [ExtensionKind.BitwiseRightShiftOperatorType]: "__luaBitwiseRightShiftBrand",
    [ExtensionKind.BitwiseRightShiftOperatorMethodType]: "__luaBitwiseRightShiftMethodBrand",
    [ExtensionKind.ConcatOperatorType]: "__luaConcatBrand",
    [ExtensionKind.ConcatOperatorMethodType]: "__luaConcatMethodBrand",
    [ExtensionKind.LessThanOperatorType]: "__luaLessThanBrand",
    [ExtensionKind.LessThanOperatorMethodType]: "__luaLessThanMethodBrand",
    [ExtensionKind.GreaterThanOperatorType]: "__luaGreaterThanBrand",
    [ExtensionKind.GreaterThanOperatorMethodType]: "__luaGreaterThanMethodBrand",
    [ExtensionKind.NegationOperatorType]: "__luaNegationBrand",
    [ExtensionKind.NegationOperatorMethodType]: "__luaNegationMethodBrand",
    [ExtensionKind.BitwiseNotOperatorType]: "__luaBitwiseNotBrand",
    [ExtensionKind.BitwiseNotOperatorMethodType]: "__luaBitwiseNotMethodBrand",
    [ExtensionKind.LengthOperatorType]: "__luaLengthBrand",
    [ExtensionKind.LengthOperatorMethodType]: "__luaLengthMethodBrand",
    [ExtensionKind.TableNewType]: "__luaTableNewBrand",
    [ExtensionKind.TableDeleteType]: "__luaTableDeleteBrand",
    [ExtensionKind.TableDeleteMethodType]: "__luaTableDeleteMethodBrand",
    [ExtensionKind.TableGetType]: "__luaTableGetBrand",
    [ExtensionKind.TableGetMethodType]: "__luaTableGetMethodBrand",
    [ExtensionKind.TableHasType]: "__luaTableHasBrand",
    [ExtensionKind.TableHasMethodType]: "__luaTableHasMethodBrand",
    [ExtensionKind.TableSetType]: "__luaTableSetBrand",
    [ExtensionKind.TableSetMethodType]: "__luaTableSetMethodBrand",
};

export function isExtensionType(type: ts.Type, extensionKind: ExtensionKind): boolean {
    const typeBrand = extensionKindToTypeBrand[extensionKind];
    return typeBrand !== undefined && type.getProperty(typeBrand) !== undefined;
}

export function isExtensionValue(
    context: TransformationContext,
    symbol: ts.Symbol,
    extensionKind: ExtensionKind
): boolean {
    return (
        symbol.getName() === extensionKindToValueName[extensionKind] &&
        symbol.declarations?.some(d => isExtensionType(context.checker.getTypeAtLocation(d), extensionKind)) === true
    );
}
