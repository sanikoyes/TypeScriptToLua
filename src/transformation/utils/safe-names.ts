import * as ts from "typescript";
import { TransformationContext } from "../context";
import { invalidAmbientIdentifierName } from "./diagnostics";
import { isSymbolExported } from "./export";
import { isAmbientNode } from "./typescript";

export const isValidLuaIdentifier = (name: string) => !luaKeywords.has(name) && /^[a-zA-Z_][a-zA-Z0-9_]*$/.test(name);
export const luaKeywords: ReadonlySet<string> = new Set([
    "and",
    "break",
    "do",
    "else",
    "elseif",
    "end",
    "false",
    "for",
    "function",
    "goto",
    "if",
    "in",
    "local",
    "nil",
    "not",
    "or",
    "repeat",
    "return",
    "then",
    "true",
    "until",
    "while",
]);

const luaBuiltins: ReadonlySet<string> = new Set([
    "_G",
    "assert",
    "coroutine",
    "debug",
    "error",
    "ipairs",
    "math",
    "pairs",
    "pcall",
    "print",
    "rawget",
    "repeat",
    "require",
    "self",
    "string",
    "table",
    "tostring",
    "type",
    "unpack",
]);

export const isUnsafeName = (name: string) => !isValidLuaIdentifier(name) || luaBuiltins.has(name);

function checkName(context: TransformationContext, name: string, node: ts.Node): boolean {
    const isInvalid = !isValidLuaIdentifier(name);

    if (isInvalid) {
        // Empty identifier is a TypeScript error
        if (name !== "") {
            context.diagnostics.push(invalidAmbientIdentifierName(node, name));
        }
    }

    return isInvalid;
}

export function hasUnsafeSymbolName(
    context: TransformationContext,
    symbol: ts.Symbol,
    tsOriginal: ts.Identifier
): boolean {
    const isAmbient = symbol.declarations?.some(d => isAmbientNode(d)) ?? false;

    // Catch ambient declarations of identifiers with bad names
    if (isAmbient && checkName(context, symbol.name, tsOriginal)) {
        return true;
    }

    // only unsafe when non-ambient and not exported
    return isUnsafeName(symbol.name) && !isAmbient && !isSymbolExported(context, symbol);
}

export function hasUnsafeIdentifierName(
    context: TransformationContext,
    identifier: ts.Identifier,
    checkSymbol = true
): boolean {
    if (checkSymbol) {
        const symbol = context.checker.getSymbolAtLocation(identifier);
        if (symbol) {
            return hasUnsafeSymbolName(context, symbol, identifier);
        }
    }

    return checkName(context, identifier.text, identifier);
}

const fixInvalidLuaIdentifier = (name: string) =>
    name.replace(/[^a-zA-Z0-9_]/g, c => `_${c.charCodeAt(0).toString(16).toUpperCase()}`);

export const createSafeName = (name: string) => "____" + fixInvalidLuaIdentifier(name);
