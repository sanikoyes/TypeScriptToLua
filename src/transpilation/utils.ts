import * as path from "path";
import * as resolve from "resolve";
import { SourceNode } from "source-map";
import * as ts from "typescript";
// TODO: Don't depend on CLI?
import * as cliDiagnostics from "../cli/diagnostics";
import * as lua from "../LuaAST";
import * as diagnosticFactories from "./diagnostics";

export interface EmitHost {
    directoryExists(path: string): boolean;
    fileExists(path: string): boolean;
    getCurrentDirectory(): string;
    readFile(path: string): string | undefined;
    writeFile: ts.WriteFileCallback;
}

interface BaseFile {
    code: string;
    sourceMap?: string;
    sourceFiles?: ts.SourceFile[];
}

export interface ProcessedFile extends BaseFile {
    fileName: string;
    luaAst?: lua.File;
    /** @internal */
    sourceMapNode?: SourceNode;
}

export interface EmitFile extends BaseFile {
    outputPath: string;
}

export const getConfigDirectory = (options: ts.CompilerOptions) =>
    options.configFilePath ? path.dirname(options.configFilePath) : process.cwd();

export function resolvePlugin(
    kind: string,
    optionName: string,
    basedir: string,
    query: string,
    importName = "default"
): { error?: ts.Diagnostic; result?: unknown } {
    if (typeof query !== "string") {
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "string") };
    }

    let resolved: string;
    try {
        resolved = resolve.sync(query, { basedir, extensions: [".js", ".ts", ".tsx"] });
    } catch (err) {
        if (err.code !== "MODULE_NOT_FOUND") throw err;
        return { error: diagnosticFactories.couldNotResolveFrom(kind, query, basedir) };
    }

    const hasNoRequireHook = require.extensions[".ts"] === undefined;
    if (hasNoRequireHook && (resolved.endsWith(".ts") || resolved.endsWith(".tsx"))) {
        try {
            const tsNodePath = resolve.sync("ts-node", { basedir });
            const tsNode: typeof import("ts-node") = require(tsNodePath);
            tsNode.register({ transpileOnly: true });
        } catch (err) {
            if (err.code !== "MODULE_NOT_FOUND") throw err;
            return { error: diagnosticFactories.toLoadItShouldBeTranspiled(kind, query) };
        }
    }

    const commonjsModule = require(resolved);
    const factoryModule = commonjsModule.__esModule ? commonjsModule : { default: commonjsModule };
    const result = factoryModule[importName];
    if (result === undefined) {
        return { error: diagnosticFactories.shouldHaveAExport(kind, query, importName) };
    }

    return { result };
}
