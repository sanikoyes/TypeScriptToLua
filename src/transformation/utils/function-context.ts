import * as ts from "typescript";
import { CompilerOptions } from "../../CompilerOptions";
import { TransformationContext } from "../context";
import { AnnotationKind, getFileAnnotations, getNodeAnnotations } from "./annotations";
import { findFirstNodeAbove, getAllCallSignatures, inferAssignedType } from "./typescript";

export enum ContextType {
    None,
    Void,
    NonVoid,
    Mixed,
}

function hasNoSelfAncestor(declaration: ts.Declaration): boolean {
    const scopeDeclaration = findFirstNodeAbove(
        declaration,
        (node): node is ts.SourceFile | ts.ModuleDeclaration => ts.isSourceFile(node) || ts.isModuleDeclaration(node)
    );

    if (!scopeDeclaration) {
        return false;
    } else if (ts.isSourceFile(scopeDeclaration)) {
        return getFileAnnotations(scopeDeclaration).has(AnnotationKind.NoSelfInFile);
    } else if (getNodeAnnotations(scopeDeclaration).has(AnnotationKind.NoSelf)) {
        return true;
    } else {
        return hasNoSelfAncestor(scopeDeclaration);
    }
}

function getExplicitThisParameter(signatureDeclaration: ts.SignatureDeclaration): ts.ParameterDeclaration | undefined {
    return signatureDeclaration.parameters.find(
        param => ts.isIdentifier(param.name) && param.name.originalKeywordKind === ts.SyntaxKind.ThisKeyword
    );
}

export function getDeclarationContextType(
    { program }: TransformationContext,
    signatureDeclaration: ts.SignatureDeclaration
): ContextType {
    const thisParameter = getExplicitThisParameter(signatureDeclaration);
    if (thisParameter) {
        // Explicit 'this'
        return thisParameter.type && thisParameter.type.kind === ts.SyntaxKind.VoidKeyword
            ? ContextType.Void
            : ContextType.NonVoid;
    }

    // noSelf declaration on function signature
    if (getNodeAnnotations(signatureDeclaration).has(AnnotationKind.NoSelf)) {
        return ContextType.Void;
    }

    if (
        ts.isMethodSignature(signatureDeclaration) ||
        ts.isMethodDeclaration(signatureDeclaration) ||
        ts.isConstructSignatureDeclaration(signatureDeclaration) ||
        ts.isConstructorDeclaration(signatureDeclaration) ||
        (signatureDeclaration.parent && ts.isPropertyDeclaration(signatureDeclaration.parent)) ||
        (signatureDeclaration.parent && ts.isPropertySignature(signatureDeclaration.parent))
    ) {
        // Class/interface methods only respect @noSelf on their parent
        const scopeDeclaration = findFirstNodeAbove(
            signatureDeclaration,
            (n): n is ts.ClassLikeDeclaration | ts.InterfaceDeclaration =>
                ts.isClassDeclaration(n) || ts.isClassExpression(n) || ts.isInterfaceDeclaration(n)
        );

        if (scopeDeclaration === undefined) {
            return ContextType.NonVoid;
        }

        if (getNodeAnnotations(scopeDeclaration).has(AnnotationKind.NoSelf)) {
            return ContextType.Void;
        }

        return ContextType.NonVoid;
    }

    // When using --noImplicitSelf and the signature is defined in a file targeted by the program apply the @noSelf rule.
    const options = program.getCompilerOptions() as CompilerOptions;
    if (options.noImplicitSelf && program.getRootFileNames().includes(signatureDeclaration.getSourceFile().fileName)) {
        return ContextType.Void;
    }

    // Walk up to find @noSelf or @noSelfInFile
    if (hasNoSelfAncestor(signatureDeclaration)) {
        return ContextType.Void;
    }

    return ContextType.NonVoid;
}

function reduceContextTypes(contexts: ContextType[]): ContextType {
    const reducer = (a: ContextType, b: ContextType) => {
        if (a === ContextType.None) {
            return b;
        } else if (b === ContextType.None) {
            return a;
        } else if (a !== b) {
            return ContextType.Mixed;
        } else {
            return a;
        }
    };

    return contexts.reduce(reducer, ContextType.None);
}

function getSignatureDeclarations(
    context: TransformationContext,
    signatures: readonly ts.Signature[]
): ts.SignatureDeclaration[] {
    return signatures.flatMap(signature => {
        const signatureDeclaration = signature.getDeclaration();
        let inferredType: ts.Type | undefined;
        if (
            ts.isMethodDeclaration(signatureDeclaration) &&
            ts.isObjectLiteralExpression(signatureDeclaration.parent) &&
            !getExplicitThisParameter(signatureDeclaration)
        ) {
            inferredType = context.checker.getContextualTypeForObjectLiteralElement(signatureDeclaration);
        } else if (
            (ts.isFunctionExpression(signatureDeclaration) || ts.isArrowFunction(signatureDeclaration)) &&
            !getExplicitThisParameter(signatureDeclaration)
        ) {
            // Infer type of function expressions/arrow functions
            inferredType = inferAssignedType(context, signatureDeclaration);
        }

        if (inferredType) {
            const inferredSignatures = getAllCallSignatures(inferredType);
            if (inferredSignatures.length > 0) {
                return inferredSignatures.map(s => s.getDeclaration());
            }
        }

        return signatureDeclaration;
    });
}

export function getFunctionContextType(context: TransformationContext, type: ts.Type): ContextType {
    if (type.isTypeParameter()) {
        type = type.getConstraint() ?? type;
    }

    if (type.isUnion()) {
        return reduceContextTypes(type.types.map(t => getFunctionContextType(context, t)));
    }

    const signatures = context.checker.getSignaturesOfType(type, ts.SignatureKind.Call);
    if (signatures.length === 0) {
        return ContextType.None;
    }

    const signatureDeclarations = getSignatureDeclarations(context, signatures);
    return reduceContextTypes(signatureDeclarations.map(s => getDeclarationContextType(context, s)));
}
