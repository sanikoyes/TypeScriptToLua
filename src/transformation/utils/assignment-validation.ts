import * as ts from "typescript";
import { getOrUpdate } from "../../utils";
import { TransformationContext } from "../context";
import {
    unsupportedNoSelfFunctionConversion,
    unsupportedOverloadAssignment,
    unsupportedSelfFunctionConversion,
} from "./diagnostics";
import { ContextType, getFunctionContextType } from "./function-context";

// TODO: Clear if types are reused between compilations
const typeValidationCache = new WeakMap<ts.Type, Set<ts.Type>>();

export function validateAssignment(
    context: TransformationContext,
    node: ts.Node,
    fromType: ts.Type,
    toType: ts.Type,
    toName?: string
): void {
    if (toType === fromType) {
        return;
    }

    if ((toType.flags & ts.TypeFlags.Any) !== 0) {
        // Assigning to un-typed variable
        return;
    }

    // Use cache to avoid repeating check for same types (protects against infinite loop in recursive types)
    const fromTypeCache = getOrUpdate(typeValidationCache, fromType, () => new Set());
    if (fromTypeCache.has(toType)) return;
    fromTypeCache.add(toType);

    validateFunctionAssignment(context, node, fromType, toType, toName);

    const fromTypeNode = context.checker.typeToTypeNode(fromType, undefined, undefined);
    const toTypeNode = context.checker.typeToTypeNode(toType, undefined, undefined);
    if (!fromTypeNode || !toTypeNode) {
        return;
    }

    if (
        (ts.isArrayTypeNode(toTypeNode) || ts.isTupleTypeNode(toTypeNode)) &&
        (ts.isArrayTypeNode(fromTypeNode) || ts.isTupleTypeNode(fromTypeNode))
    ) {
        // Recurse into arrays/tuples
        const fromTypeArguments = (fromType as ts.TypeReference).typeArguments;
        const toTypeArguments = (toType as ts.TypeReference).typeArguments;

        if (fromTypeArguments === undefined || toTypeArguments === undefined) {
            return;
        }

        const count = Math.min(fromTypeArguments.length, toTypeArguments.length);
        for (let i = 0; i < count; ++i) {
            validateAssignment(context, node, fromTypeArguments[i], toTypeArguments[i], toName);
        }
    }

    if (
        (toType.flags & ts.TypeFlags.Object) !== 0 &&
        (ts.isTypeLiteralNode(toTypeNode) ||
            ((toType as ts.ObjectType).objectFlags & ts.ObjectFlags.ClassOrInterface) !== 0) &&
        toType.symbol &&
        toType.symbol.members &&
        fromType.symbol &&
        fromType.symbol.members
    ) {
        // Recurse into interfaces
        toType.symbol.members.forEach((toMember, escapedMemberName) => {
            if (fromType.symbol.members) {
                const fromMember = fromType.symbol.members.get(escapedMemberName);
                if (fromMember) {
                    const toMemberType = context.checker.getTypeOfSymbolAtLocation(toMember, node);
                    const fromMemberType = context.checker.getTypeOfSymbolAtLocation(fromMember, node);
                    const memberName = ts.unescapeLeadingUnderscores(escapedMemberName);
                    validateAssignment(
                        context,
                        node,
                        fromMemberType,
                        toMemberType,
                        toName ? `${toName}.${memberName}` : memberName
                    );
                }
            }
        });
    }
}

function validateFunctionAssignment(
    context: TransformationContext,
    node: ts.Node,
    fromType: ts.Type,
    toType: ts.Type,
    toName?: string
): void {
    const fromContext = getFunctionContextType(context, fromType);
    const toContext = getFunctionContextType(context, toType);

    if (fromContext === ContextType.Mixed || toContext === ContextType.Mixed) {
        context.diagnostics.push(unsupportedOverloadAssignment(node, toName));
    } else if (fromContext !== toContext && fromContext !== ContextType.None && toContext !== ContextType.None) {
        context.diagnostics.push(
            toContext === ContextType.Void
                ? unsupportedNoSelfFunctionConversion(node, toName)
                : unsupportedSelfFunctionConversion(node, toName)
        );
    }
}
