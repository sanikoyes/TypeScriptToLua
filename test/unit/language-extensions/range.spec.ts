import * as util from "../../util";
import { invalidRangeControlVariable, invalidRangeUse } from "../../../src/transformation/utils/diagnostics";

test("$range basic use", () => {
    util.testFunction`
        const result: number[] = [];
        for (const i of $range(1, 5)) {
            result.push(i);
        }
        return result;
    `
        .withLanguageExtensions()
        .expectToEqual([1, 2, 3, 4, 5]);
});

test("$range basic use with step", () => {
    util.testFunction`
        const result: number[] = [];
        for (const i of $range(1, 10, 2)) {
            result.push(i);
        }
        return result;
    `
        .withLanguageExtensions()
        .expectToEqual([1, 3, 5, 7, 9]);
});

test("$range basic use reverse", () => {
    util.testFunction`
        const result: number[] = [];
        for (const i of $range(5, 1, -1)) {
            result.push(i);
        }
        return result;
    `
        .withLanguageExtensions()
        .expectToEqual([5, 4, 3, 2, 1]);
});

test("$range invalid control variable", () => {
    util.testFunction`
        let i: number;
        for (i of $range(1, 5)) {}
    `
        .withLanguageExtensions()
        .expectDiagnosticsToMatchSnapshot([invalidRangeControlVariable.code]);
});

test.each([
    "for (const i in $range(1, 10, 2)) {}",
    "const x = $range(1, 10);",
    "const range = $range;",
    "const y = [...$range(1, 10)];",
    "for (const i of ($range(1, 10, 2))) {}",
    "for (const i of $range(1, 10, 2) as number[]) {}",
])("$range invalid use (%p)", statement => {
    util.testFunction`
        ${statement}
    `
        .withLanguageExtensions()
        .expectDiagnosticsToMatchSnapshot([invalidRangeUse.code]);
});
