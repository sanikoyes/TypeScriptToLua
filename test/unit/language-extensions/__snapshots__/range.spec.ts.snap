// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$range invalid control variable: code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local i
    for ____ = 1, 5 do
    end
end
return ____exports"
`;

exports[`$range invalid control variable: diagnostics 1`] = `"main.ts(3,14): error TSTL: For loop using $range must declare a single control variable."`;

exports[`$range invalid use ("const range = $range;"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local range = ____
end
return ____exports"
`;

exports[`$range invalid use ("const range = $range;"): diagnostics 1`] = `"main.ts(2,23): error TSTL: $range can only be used in a for...of loop."`;

exports[`$range invalid use ("const x = $range(1, 10);"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local x = ____(nil, 1, 10)
end
return ____exports"
`;

exports[`$range invalid use ("const x = $range(1, 10);"): diagnostics 1`] = `"main.ts(2,19): error TSTL: $range can only be used in a for...of loop."`;

exports[`$range invalid use ("const y = [...$range(1, 10)];"): code 1`] = `
"local lualib_bundle = require(\\"lualib_bundle\\");
local _ENV = setmetatable({}, { __index = lualib_bundle })
pcall(function() setfenv(1, _ENV) end)
local ____exports = {}
function ____exports.__main(self)
    local y = {
        __TS__Spread(
            ____(nil, 1, 10)
        )
    }
end
return ____exports"
`;

exports[`$range invalid use ("const y = [...$range(1, 10)];"): diagnostics 1`] = `"main.ts(2,23): error TSTL: $range can only be used in a for...of loop."`;

exports[`$range invalid use ("for (const i in $range(1, 10, 2)) {}"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    for i in pairs(
        ____(nil, 1, 10, 2)
    ) do
    end
end
return ____exports"
`;

exports[`$range invalid use ("for (const i in $range(1, 10, 2)) {}"): diagnostics 1`] = `"main.ts(2,25): error TSTL: $range can only be used in a for...of loop."`;

exports[`$range invalid use ("for (const i of $range(1, 10, 2) as number[]) {}"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    for ____, i in ipairs(
        ____(nil, 1, 10, 2)
    ) do
    end
end
return ____exports"
`;

exports[`$range invalid use ("for (const i of $range(1, 10, 2) as number[]) {}"): diagnostics 1`] = `"main.ts(2,25): error TSTL: $range can only be used in a for...of loop."`;

exports[`$range invalid use ("for (const i of ($range(1, 10, 2))) {}"): code 1`] = `
"local lualib_bundle = require(\\"lualib_bundle\\");
local _ENV = setmetatable({}, { __index = lualib_bundle })
pcall(function() setfenv(1, _ENV) end)
local ____exports = {}
function ____exports.__main(self)
    for ____, i in __TS__Iterator(
        ____(nil, 1, 10, 2)
    ) do
    end
end
return ____exports"
`;

exports[`$range invalid use ("for (const i of ($range(1, 10, 2))) {}"): diagnostics 1`] = `"main.ts(2,26): error TSTL: $range can only be used in a for...of loop."`;
